File: 068.rb
# 068
# Text Justification
# 2015-12-22
#####################################################
# @param {String[]} words
# @param {Integer} max_width
# @return {String[]}
def full_justify(words, max_width)
  v, line = [], []
  words.each do |word|
    if (line+[word]).join(' ').size > max_width
      v.push(join_line(line, max_width))
      line = []
    end
    line.push(word)
  end
  v.push(join_line(line, max_width, true))
end

def join_line(line, max_width, last = false)
  return line.join(' ') + ' '*(max_width-line.join(' ').size) if (line.size == 1 or last == true)
  space, extra = (max_width-line.join.size).divmod(line.size-1)
  line[0..extra].join(' '*(space+1)) + ' '*space + line[extra+1..-1].join(' '*space)
end

#####################################################

words = ["This", "is", "an", "example", "of", "text", "justification."]
max_width = 16
r = full_justify(words, max_width)
r.each do |s|
  puts "\"#{s}\""
end

File: 068.py
#   068
#   Test Justification
#   2015-12-22
#####################################################
class Solution(object):
    def fullJustify(self, words, maxWidth):
        """
        :type words: List[str]
        :type maxWidth: int
        :rtype: List[str]
        """
        v, line = [], []
        for word in words:
            if len(' '.join(line+[word])) > maxWidth:
                v.append(self.joinLine(line, maxWidth))
                line = []
            line.append(word)
        v.append(self.joinLine(line, maxWidth, True))
        return v

    def joinLine(self, line, maxWidth, last = False):
        if (len(line) == 1 or last == True):
            return ' '.join(line) + ' '*(maxWidth - len(' '.join(line)))
        space, extra = divmod(maxWidth - len(''.join(line)), len(line)-1)
        return (' '*(space+1)).join(line[:extra+1]) + ' '*space + (' '*space).join(line[extra+1:])

#####################################################
sol = Solution()

words = ["This", "is", "an", "example", "of", "text", "justification."]
maxWidth = 16
r = sol.fullJustify(words, maxWidth)

for word in r:
    print '"'+word+'"'

